gaphor/storage/storage.py

```diff
@@ -462,3 +469,135 @@ def upgrade_note_on_model_element_only(
+        min_y = min((y for _, y in points), default=0)
+        max_x = max((x for x, _ in points), default=min_x)
+        max_y = max((y for _, y in points), default=min_y)
+
+        return (min_x, min_y, max_x - min_x, max_y - min_y)
+
+    def add_frame(diagram):
+        # Create frame
+        frame = element(generate_id(), "DiagramFrameItem")
+        elements[frame.id] = frame
+        # Attach subject
+        if element_id := diagram.references.get("element", None):
+            subject = elements[element_id]
+            frame.references["subject"] = subject.id
+            subject.references.setdefault("presentation", []).append(frame.id)
+        # Position and size the frame
+        min_x, min_y, width, height = parse_diagram_boundaries(diagram)
+        padding = 50
+        frame.values[
+            "matrix"
+        ] = f"(1.0, 0.0, 0.0, 1.0, {min_x - padding}, {min_y - padding})"
+        frame.values["top-left"] = "(0.0, 0.0)"
+        frame.values["width"] = f"{width + 2 * padding}"
+        frame.values["height"] = f"{height + 2 * padding}"
+        # Group the parentless items to the frame
+        parentless_items = (
+            item
+            for item in (
+                elements[id] for id in diagram.references.get("ownedPresentation", [])
+            )
+            if not item.references.get("parent", None)
+        )
+        for item in parentless_items:
+            frame.references.setdefault("children", []).append(item.id)  # type: ignore[union-attr] # we know for a fact that it can only be list, because we set it here first time
+            item.references["parent"] = frame.id
+            (a, b, c, d, x, y) = literal_eval(item.values.get("matrix"))
+            # TODO: use matrix math here
+            item.values[
+                "matrix"
+            ] = f"({a}, {b}, {c}, {d}, {x - (min_x - padding)}, {y - (min_y - padding)})"
+        # Attach the frame to the diagram
+        frame.references["diagram"] = diagram.id
+        diagram.references.setdefault("ownedPresentation", []).append(frame.id)
+
+    def replace_diagram_element_item_with_frame(item):
+        item.type = "DiagramFrameItem"
+
+    for diagram in diagrams:
+        if any(
+            elements[id].type == "DiagramFrameItem"
+            for id in diagram.references.get("ownedPresentation", [])
+        ):
+            continue
+
+        if el := diagram.references.get("element", None):
+            element_items = [
+                item
+                for item in (
+                    elements[id]
+                    for id in diagram.references.get("ownedPresentation", [])
+                )
+                if item.references.get("subject", None) == el
+            ]
+
+            if len(element_items) == 1:
+                replace_diagram_element_item_with_frame(element_items[0])
+            else:
+                add_frame(diagram)
+        else:
+            add_frame(diagram)
```

Comment:
Are these acceptable defaults?

It ensures that there is always a frame which surrounds all the existing diagram items. If there is a single item which presents the same element as the diagram itself; we replace it with frame instead.

The latter may not be the safest thing to do, especially if we have a bdd representing a block; and we have the same block on the diagram which has the associations with other blocks; it may either: a) look like we form the associations with a diagram; b) the associations visualization may be broken. I suppose the option (a). But it is not desirable substitution in this case, so I am also considering just adding frame and surrounding all the items and call it a day and leave the cleanup for the modeler.

I tested this with my models; and on all of them; it was as I expected, but I can be considered rather strict in modeling and I do not use the block to define the block's internals as it would be kinda like defining systems engineering as "Systems Engineering is something that Systems Engineers do" where Systems Engineer is defined as "Someone who does Systems Engineering".

Again, I am interested in having a discussions.

Reply:
I've decided to play it save an simply add a diagram frame item on all the sysml diagrams without any fancy magic. I leave it up to the modeler to decide whether they want to continue modeling the old ways; or they like to leverage the frame's ability to hold ports/parameters. I wont enforce this. :slightly_smiling_face: